### PR TITLE
Passivation configured in language supports via the discovery protocol

### LINKED ISSUE #298: Update entity passivation
Akka 2.6 has entity passivation built-in now. See `akka.cluster.sharding.passivate-idle-entity-after` setting and [discussion](https://github.com/cloudstateio/cloudstate/pull/293/files#r419180864).

Or we may want to allow passivation timeouts to be configurable per entity type on the user side?

### PR DESCRIPTION
**Pull request by ralphlaude**
Resolves #298

Passivation timeouts configured in language supports via the discovery protocol. The first step is the java-support and then node-support.

In the java-support all entity annotations are extended `@XXXEntity (passivationTimeout = 10)`,  the service class `io.cloudstate.javasupport.Service` is also extented with `passivationTimeout` and the grpc entity in the entity.proto.

### CODE REVIEW

**Comment by ralphlaude** [comment]
@pvlugter what do you think about the option?

File: java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java
Comment (sleipnir) [inline java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Hello Guy I didn't like this option for me it should be done at the Entity registration stage and not at the entity itself.

**Review by sleipnir — COMMENTED** [review]
Guy thinks that this option should be passed via registryXXXXEntity methods and not in the entity's annotation. I think this pollutes the entity a little.

File: java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java
Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Hi Adriano, it is also fine for me and it is easier. I will do it at the registration stage.

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Great! Thanks

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: why do you like this option?

Comment (marcellanz) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Why would 0 not be a valid value? And therefore passivation disabled at all?

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Doing it at the registration stage means breaking the existing registration API and all stay the same with annotations. Then we are forcing user to think about the passivation timeout for the registration. I think the option with annotations is without breaking and easier for the user. @sleipnir I would be glad to hear your idea.

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > why do you like this option?

I prefer the registration option because I understand a registration step as a configuration step for the execution service of the entity and not a property of the entity. I also prefer that the entity annotations are as simple as possible, without many options because it seems cleaner. See for example how I make the settings in Kotlin support, the user informs what he wants in the registration step and the entity remains without many changes.

```
fun main() {
    cloudstate {
        eventsourced {
            entityService = ShoppingCartEntity::class
            descriptor = Shoppingcart.getDescriptor().findServiceByName("ShoppingCart")
            additionalDescriptors = mutableListOf(Shoppingcart.getDescriptor(), Domain.getDescriptor() )
           
            // Settings that I want to inform the proxy are here:
            snapshotEvery = 1
            persistenceId = "shopping-cart"
        }

    }.start()
            .toCompletableFuture()
            .get()
}
```

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: @marcellanz 0 or negative means use the default config for the passivation.

Comment (marcellanz) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: @ralphlaude yeah saw that documented, but what is the motivation to allow a low time of 1sec, but not 0sec? <0 would be enough to piggyback to take the default config, alas setting to 0 allows the entity to disable or set it to 0.

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > I prefer the registration option because I understand a registration step as a configuration step for the execution service of the entity and not a property of the entity. I also prefer that the entity annotations are as simple as possible, without many options because it seems cleaner. See for example how I make the settings in Kotlin support, the user informs what he wants in the registration step and the entity remains without many changes.

@sleipnir it is just about Java-support here and other languages can use their own ways to do that. I have not idea about Kotlin but I don't see another option in Java. How would you configure the passivation timeout with Kotlin?

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > Doing it at the registration stage means breaking the existing registration API and all stay the same with annotations. Then we are forcing user to think about the passivation timeout for the registration. I think the option with annotations is without breaking and easier for the user. @sleipnir I would be glad to hear your idea.

I do not agree. I don't see how adding an option to the registry breaks the current API? Do the values not have defaults? Am I required to set the passivation time? If any of the answers to these questions are true then it would break the registration API, otherwise it would not.

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: @ralphlaude If the user for any reason wants to use the passivation option then he is already thinking about Passivation, he is not being forced to do anything whether it be done in the annotation or in the registry phase.

Comment (marcellanz) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: I should have written instead of
> setting to 0 allows the entity to disable or set it to 0

"setting to 0 allows the entity passivate immediately."

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > > I prefer the registration option because I understand a registration step as a configuration step for the execution service of the entity and not a property of the entity. I also prefer that the entity annotations are as simple as possible, without many options because it seems cleaner. See for example how I make the settings in Kotlin support, the user informs what he wants in the registration step and the entity remains without many changes.
> 
> @sleipnir it is just about Java-support here and other languages can use their own ways to do that. I have not idea about Kotlin but I don't see another option in Java. How would you configure the passivation timeout with Kotlin?

I know that we are talking about java support, I just gave you an example to explain the concept of what I was telling you. If you don't think it is necessary to change then ok, I just presented my arguments.

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: @marcellanz the entity should not passivate immediately and it should be allowed. Setting to 0 allows the entity to passivate after 30 seconds. The`passivateTimeout` in the annotation should define the minimal value to accept.

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > @marcellanz the entity should not passivate immediately and it should be allowed. Setting to 0 allows the entity to passivate after 30 seconds. The`passivateTimeout` in the annotation should define the minimal value to accept.

This seems counter intuitive to me and should be well documented if it is assumed. Setting a property to 0 and assuming a value of 30 is somewhat unusual. If this is really the case, we cannot forget to be very clear about this in the documentation.

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: @sleipnir passivation is always enabled and is implicit to the user. The `registerXXXEntity` does not  force the user to think about passivation. The new `registerXXXEntity(..., passivationTimeout)` forces the user to provide passivation timeout or at least to think about it.
For now i see two options here:
1- Annotations
2- Provide overloaded registration methods with passivationTimeout

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > @sleipnir passivation is always enabled and is implicit to the user. The `registerXXXEntity` does not force the user to think about passivation. The new `registerXXXEntity(..., passivationTimeout)` forces the user to provide passivation timeout or at least to think about it.
> For now i see two options here:
> 1- Annotations
> 2- Provide overloaded registration methods with passivationTimeout

Guy Registry phase doesn't force anyone, I already told you, sorry your argument doesn't convince me, I'm sorry. As I told you before, if you don't feel comfortable don't change, I made my contribution to the topic and I will not argue anymore.

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > I know that we are talking about java support, I just gave you an example to explain the concept of what I was telling you. If you don't think it is necessary to change then ok, I just presented my arguments.

understood and i saw how it works in Kotlin

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: > This seems counter intuitive to me and should be well documented if it is assumed. Setting a property to 0 and assuming a value of 30 is somewhat unusual. If this is really the case, we cannot forget to be very clear about this in the documentation.

@sleipnir I agree with default values in annotations and it already works that way with attributes (persistentId and snapshotEvery) in `@XXXEntity`. It is also well documented. May be we should change that and explicitly work with default values in annotations.

Comment (pvlugter) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: I agree generally about keeping the entity implementations simpler by not having too many options on the annotations. Also understand that this is the current approach in Java support, and annotation settings allow for defaults. So I'd be inclined to have both annotation and registration options right now. The registration method for annotated entity classes doesn't currently have extra options, but it makes sense to add them now. Not sure how many more options there will be in the future, but I'd consider having a settings object (so something like `EventSourcedOptions` or similar), instead of supporting multiple permutations of parameters — so the method can just be overloaded with and without the options, and so that the options object can have default or unset options.

Regarding default of 0, I'd probably interpret a timeout of 0 as disabling passivation altogether. Not sure if we would want to support disabling passivation, where the user function would control lifecycle itself or something, but I think I'd avoid 0 as being used for an unset option for selecting the proxy default.

I think it makes sense to improve the APIs and protocol approaches whenever possible. For a timeout setting, it would be good to support any kind of duration (not just int of seconds). So I would consider accepting `java.time.Duration` in the java support APIs, or `TimeUnit` variations. Annotation values are limited, can't use arbitrary classes, so one option instead of an int is to use a String and use the typesafe Config duration parsing or `Duration.parse` (which is a bit different). And can then use empty String as having the timeout unset.

For the settings that we want to have fall back to proxy-side defaults, I would support this directly in the settings — so the entity options object knows when a timeout has been set or not, and same in the protocol — and it only sends a timeout if it's been configured in the user function, rather than relying on negative numbers. And the protocol could have a duration message, with time units, to make things complete.

So I would aim for something along the lines of:

```java
// no passivation timeout set, nothing sent with entity spec to proxy
@EventSourcedEntity(persistenceId = "something")
```

```java
// passivation timeout parsed from string, sent as duration message in protocol
@EventSourcedEntity(persistenceId = "something", passivationTimeout = "10s")
```

```java
// no extra options specified, use annotation options or defaults
.registerEventSourcedEntity(
  SomeEntity.class,
  SomeProtocol.getDescriptor().findServiceByName("SomeService"),
  SomeProtocol.getDescriptor())
```

```java
// options specified, which will override annotation options or defaults
.registerEventSourcedEntity(
  SomeEntity.class,
  SomeProtocol.getDescriptor().findServiceByName("SomeService"),
  SomeProtocol.getDescriptor(),
  EventSourcedOptions.defaults().setPassivationTimeout(Duration.ofSeconds(10))
```

Comment (sleipnir) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: I agree with the whole proposal here

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: After the discussion last week, i agree fully with the proposal.

Comment (marcellanz) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: >and it only sends a timeout if it's been configured in the user function, rather than relying on negative numbers. And the protocol could have a duration message, with time units, to make things complete.

I agree to change the protocol now and a soon as possible for such changes. I also agree on not depending on magic values.

Comment (jroper) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: I think I'd like to have options for expression the following things - perhaps not immediately, but we should choose a mechanism and protocol that allows us to add support for them in future if needed:

* A specific passivation timeout, with at least millisecond resolution.
* Default - note that I think it's fine for the default to come from the support library, I don't see any reason why the proxy needs to be in control of the default.
* No passivation timeout - entities should never passivate unless rebalanced.
* Immediate passivation, this effectively means the state is reloaded for each command.
* Some way to indicate the proxy should apply smarter, more dynamic passivation strategies, for example strategies that respond to resource constraints, or perhaps auto-tune themselves based on cardinality and/or hot spot metrics.

So, basically, I think it's worth now choosing a strategy in the APIs and protocol that will allow us to add some of the features above without breaking the protocol or APIs.

Comment (pvlugter) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Good idea to future-proof the protocol and APIs now for other passivation options.

Comment (ralphlaude) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: Really good idea for passivation options.
More generally we should have a strategy for adding other future options (not only passivation) without breaking the protocol or the APIs.

Comment (pvlugter) [reply java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java:50]: After some more discussion:

- passivation timeout in protocol can just be milliseconds, we don't expect more fine-grained than that
- language supports can use duration APIs or similar as appropriate, convert to milliseconds
- protocol should support expanding to other passivation strategies, where initially it could just be `timeout`, ideally we can add `never` or `immediate` or other strategies in a protocol-compatible way
- protocol should support strategies explicitly, such as `timeout`, `never` or `immediate`, and not use magic values

**Comment by ralphlaude** [comment]
The discovery process, the proxy and the API have been extended to support passivation strategy and also that new passivation strategy (`never`, `immedaite` and perhaps resource specifics strategies) can be added without breaking the API. The only supported passivation strategy right now is `timeout`. Passivation strategy could be as complex as possible to configure and this does not fit within annotations. So the approach to pass configuration parameter at registration time is a better fit. There is `EntityOptions` for configuring passivation strategy and other aspects if needed.
The language support should deal with default passivation timeout and more generally fallback.
All entities must support a passavation strategy and right now it is `timeout`. The entity.proto file has been modified like this:
```
message Entity {
    ...

    // The passivation strategy for the entity.
    EntityPassivationStrategy passivation_strategy = 4;
}

// A passivation strategy for the entity which is sent to the proxy.
message EntityPassivationStrategy {
    oneof strategy {
	// the timeout passivation strategy.
        TimeoutPassivationStrategy timeout = 1;
        ImmediatePassivationStrategy immediate = 2; // this strategy can be added later on
        NeverPassivationStrategy never = 3; // this strategy can also be added later on
        // more strategies can be added here
    }
}

// A passivation strategy based on a timeout. The idle timeout after which an entity is passivated.
message TimeoutPassivationStrategy {
    // The timeout in millis
    int64 timeout = 1;
}

// A passivation strategy which passivates the entity after a command is handled.
message ImmediatePassivationStrategy {}

// A passivation strategy which never passivates the entity.
message NeverPassivationStrategy {}
```

For the java support it will be like at the end:
```java

// passivation stratgey timeout with default passivation timeout set which is sent with entity spec to proxy
@EventSourcedEntity(persistenceId = "something")

// no extra options specified, passivation stratgey timeout with default timeout sent with entity spec to proxy
.registerEventSourcedEntity(
  SomeEntity.class,
  SomeProtocol.getDescriptor().findServiceByName("SomeService"),
  SomeProtocol.getDescriptor())

// options specified with passivation stratgey timeout, which will be sent with entity spec to proxy
.registerEventSourcedEntity(
  SomeEntity.class,
  SomeProtocol.getDescriptor().findServiceByName("SomeService"),
  EventSourcedEntityOptions.defaults().withPassivationStrategy(PassivationStrategy.timeout(Duration.ofSeconds(10)))
  SomeProtocol.getDescriptor())
```

Only the java support is implemented and all others language supports (Nodejs, GO, Kotlin, Python, Spring and so on) should be extended before this can be merged.
Wo can help with other language supports (Nodejs, GO, Kotlin, Python, Spring and so on)?

**Comment by marcellanz** [comment]
Thanks @ralphlaude :) For Go for sure I can help.
If needed we could work based on a dedicated branch so that maintained language supports can be aligned. As the new field `EntityPassivationStrategy` protocol-wise is an addition, entity discovery could probably even work without the support on user languages?

**Comment by ralphlaude** [comment]
@marcellanz thanks for Go.
Yes we can think how to work on a dedicated branch to be aligned.
With the addition of the new field `EntityPassivationStrategy` in the protocol the discovery is broken without the support on user language. It can be changed if we want it.

**Comment by sleipnir** [comment]
Hi @ralphlaude , thanks for another one!
I can help with the other languages mentioned but generally as Kotlin and Springboot depend on the java-support jar I choose to wait for the release of an official release with the functionality, I do this to not have to build the jars locally. I sincerely wish there were more releases to help me with this, but I may have to follow the branches of the main repository with equivalent branches in the languages, I don't know, I find this approach very laborious, but this should be discussed in another forum.

**Comment by marcellanz** [comment]
@sleipnir I think, it should be possible or better said desirable for language supports to evolve before the core project releases. For non-JVM supports it is already manageable with tck- and proxy images available from master. For JVM support perhaps snapshot releases help? But as you said, could be discussed elsewhere.

**Comment by sleipnir** [comment]
> For JVM support perhaps snapshot releases help? ...

Yes!

**Comment by pvlugter** [comment]
> > For JVM support perhaps snapshot releases help? ...
> 
> Yes! But they are also often not available.

It's easy to publish a versioned snapshot for java support. Just ask if you need one :)

**Comment by pvlugter** [comment]
> With the addition of the new field `EntityPassivationStrategy` in the protocol the discovery is broken without the support on user language. It can be changed if we want it.

Since it's an additional field, the protocol can be compatible with older versions. In the proxy, instead of throwing exceptions if the language support doesn't define a passivation strategy, we can use the proxy-side default.

**Comment by ralphlaude** [comment]
> Since it's an additional field, the protocol can be compatible with older versions. In the proxy, instead of throwing exceptions if the language support doesn't define a passivation strategy, we can use the proxy-side default.
> 
> I think it will be better to have this compatible and use the proxy default, rather than require all language supports to be updated straight away. There's also an older issue on configurable entity passivation (#111) which describes having the passivation strategy defined in the Kubernetes CRD or ConfigMap and injected by the operator — which would allow the proxy-side default to be configured, if the language support doesn't define it.

The proxy-default for passivation strategy will be introduced again to allow the compatibility so we can move forward. I think it would better in the middle term to have a central place where to configure things like passivation strategy. Now we have two places where passivation strategy is configured. I think the better place for that is the language support.

**Comment by ralphlaude** [comment]
Everything is done here. I am not sure if we want to extends the nodejs support here or do it in another issue. If yes, I will create issues for nodejs support and another language support (GO, Spring, Kotlin, Python and others).
@sleipnir, @marcellanz what do you think?

**Comment by marcellanz** [comment]
@ralphlaude I think for changes on the protocol one implementation for a "main" user support language should be enough. With "main" this might be JS or Java I think.

As the protocol changed more often lately, it reveals how we might work with such changes in general. With changes worked on concurrently, having the proxy and TCK in the right state for development leads to a bit of coordination. 
I think we don't need overly complex rules for that like, as an example, opening issues for all support language. Perhaps the TCK can help to guide where changes are pending for other support languages. This spans a matrix of combinations for proxy, TCK and support language implementations. The current situation already shows this, with currently a few languages aligned with master, and others partly up to date.

This said, I think support languages should aim to get protocol updates by themselves. If the protocol changes by a PR like this one, the changing PR a) should provide a TCK test or if not applicable, b) a comment in the spec about semantics changed in the protocol.

The TCK Verification here https://observablehq.com/@marcellanz/cloudstate-tck-verification displays the current situation also in an automatic way, with pending TCK tests visible.

WDYT?

**Comment by sleipnir** [comment]
@ralphlaude  I agree with @marcellanz  but I add that the comment on the semantics desired by the protocol must be mandatory in PR. This avoids that developers have to have a long journey of understanding evaluating codes and contracts where a comment would be enough to make the intention clear.

**Comment by ralphlaude** [comment]
@marcellanz the proposal is fine for me. It would be nice to have support for nodejs. I will comment the spec about the new semantic.

**Comment by ralphlaude** [comment]
@pvlugter, @jroper I need support here for nodejs. Please take a look.

File: java-support/src/main/scala/io/cloudstate/javasupport/impl/PassivationStrategies.scala
Comment (pvlugter) [inline java-support/src/main/scala/io/cloudstate/javasupport/impl/PassivationStrategies.scala:29]: Ideally the config is not loaded in multiple places (even though the config loading is cached), so that it's only defined once, especially if custom config can be passed into the top-level classes. Looks like `CloudStateRunner` has a constructor that accepts a config. Could we get move the defaults to one place, so they can all use the same config?

File: protocols/protocol/cloudstate/entity.proto
Comment (pvlugter) [inline protocols/protocol/cloudstate/entity.proto:274]:
```suggestion
// If an entity user function does not configure the passivation strategy the proxy uses its fallback default value.
```

File: protocols/protocol/cloudstate/entity.proto
Comment (pvlugter) [inline protocols/protocol/cloudstate/entity.proto:284]:
```suggestion
// A passivation strategy based on a timeout. The idle timeout after which a user function's entity is passivated.
```

File: proxy/core/src/main/scala/io/cloudstate/proxy/crdt/CrdtSupportFactory.scala
Comment (pvlugter) [inline proxy/core/src/main/scala/io/cloudstate/proxy/crdt/CrdtSupportFactory.scala:100]: Is this to fix the string interpolation missing from the second string? Rather than multiline string, with margin stripped, and then newlines replaced, could just add string interpolation to the second string (`s"`) and not the first one.

File: proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedSupportFactory.scala
Comment (pvlugter) [inline proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedSupportFactory.scala:92]: Same here, if it's to fix the missing string interpolation, just move the `s"` to the second string.

**Review by pvlugter — COMMENTED** [review]
Looking good. Added a few review comments — would be good to not load the config in the Timeout static object in particular, but have the default come from the top-level config.

**Comment by pvlugter** [comment]
Yes, would be good to add passivation timeouts for node support. Node support is behind the Java support in a few areas now: applying events immediately (and any failure handling for this) #375, support for value entities, new TCK implementation, and this. Doesn't need to be added to this PR, let's create an issue to track.

**Comment by pvlugter** [comment]
Could be useful to have passivation timeouts as part of the TCK for testing.

File: java-support/src/main/scala/io/cloudstate/javasupport/impl/PassivationStrategies.scala
Comment (ralphlaude) [reply java-support/src/main/scala/io/cloudstate/javasupport/impl/PassivationStrategies.scala:29]: I will take a look and see how the issue can be solved.

Comment (ralphlaude) [reply java-support/src/main/scala/io/cloudstate/javasupport/impl/PassivationStrategies.scala:29]: > Could we get move the defaults to one place, so they can all use the same config?

@pvlugter I am not sure I understand. Could you please be more specific?

**Comment by ralphlaude** [comment]
> Could be useful to have passivation timeouts as part of the TCK for testing.

@pvlugter Could you please be more specific?

File: java-support/src/main/scala/io/cloudstate/javasupport/CloudStateRunner.scala
Comment (pvlugter) [inline java-support/src/main/scala/io/cloudstate/javasupport/CloudStateRunner.scala:86]: It's the other constructor, that takes a config explicitly, that should be supported. The config gets passed to the actor system. It should also be the config that gets passed through to other places (could be retrieved from the actor system). Might be more obvious if there's a test that sets the passivation timeout in config?

**Comment by pvlugter** [comment]
> > Could be useful to have passivation timeouts as part of the TCK for testing.
> 
> @pvlugter Could you please be more specific?

Since passivation strategy is part of the protocol, it would be useful to test it in the TCK. So testing that language supports can configure this, and ideally also testing that the proxy applies the timeout and passivates the entity. For the TCK, it would be fine to have an additional entity defined which is just for testing this, since the passivation timeout is on entity discovery.

File: java-support/src/main/scala/io/cloudstate/javasupport/CloudStateRunner.scala
Comment (ralphlaude) [reply java-support/src/main/scala/io/cloudstate/javasupport/CloudStateRunner.scala]: @pvlugter I hope my solution does the job and I did not find a better way.

**Comment by ralphlaude** [comment]
TCK tests for passivation has been added which uses value-based entities.

File: tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala
Comment (pvlugter) [inline tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala:183]: Don't think it makes sense to include a check for none or not none :)

File: tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala
Comment (pvlugter) [inline tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala:236]: May need to be longer for CI tests

**Review by pvlugter — COMMENTED** [review]
Okay, I can always follow up with something that doesn't use a global static config

File: tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala
Comment (marcellanz) [inline tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala:198]: This would then show this test "green" for language supports not implementing passivation Strategy? 

If yes, perhaps this should show a failed test. The TCK should show which language support libraries are up-to-date. The TCK will probably show for some time a lot of "red and orange" for other implementations
https://observablehq.com/@marcellanz/cloudstate-tck-verification as the protocol progresses.

File: proxy/core/src/main/resources/reference.conf
Comment (marcellanz) [inline proxy/core/src/main/resources/reference.conf:150]: Are 30s then also a good option for production?
Perhaps we might document that, that the default is not the value to be chosen for production environments.

Comment (ralphlaude) [reply proxy/core/src/main/resources/reference.conf:150]: It is true and it will be documented. I am not sure if 30s is a good option for production. Perhaps @pvlugter can give more insight here.

File: tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala
Comment (ralphlaude) [reply tck/src/main/scala/io/cloudstate/tck/CloudStateTCK.scala:198]: Good point. I will rather try to put the test on pending for language support not implementing the passivation strategy

**Comment by ralphlaude** [comment]
@pvlugter ping! you can follow up.

**Comment by pvlugter** [comment]
I'll just close and reopen this PR to see if we have Travis CI working again.

**Review by pvlugter — APPROVED** [review]
👍🏼 Good to see the global config part updated.

I'll merge this now, and include in the TCK reorganisation.

**Pull request merged**
